pkg/pool-hooks/contracts/StableSurgeHookExample.sol

```diff
@@ -0,0 +1,243 @@
+    }
+
+    /**
+     * @notice Sets the hook threshold percentage.
+     * @dev This function must be permissioned. If the pool does not have a swap fee manager role set, the surge
+     * threshold will be effectively immutable, set to the default threshold for this hook contract.
+     */
+    function setSurgeThresholdPercentage(address pool, uint256 newSurgeThresholdPercentage) external {
+        if (_vault.getPoolRoleAccounts(pool).swapFeeManager != msg.sender) {
+            revert SenderNotAllowed();
+        }
+        _ensureValidPercentage(newSurgeThresholdPercentage);
+
+        _setSurgeThresholdPercentage(pool, newSurgeThresholdPercentage);
+    }
+
+    /**
+     * @notice Calculate the surge fee percentage. If below threshold, return the standard static swap fee percentage.
+     * @dev It is public to allow it to be called off-chain.
+     * @param params Input parameters for the swap (balances needed)
+     * @param amountCalculatedScaled18 THe post-swap amountCalculated, assuming the standard fee is charged
+     * @param surgeThresholdPercentage The current surge threshold percentage for this pool
+     * @param staticFeePercentage The static fee percentage for the pool (default if there is no surge)
+     */
+    function getSurgeFeePercentage(
+        PoolSwapParams calldata params,
+        uint256 amountCalculatedScaled18,
+        uint256 surgeThresholdPercentage,
+        uint256 staticFeePercentage
+    ) public pure returns (uint256) {
+        uint256 numTokens = params.balancesScaled18.length;
+
+        uint256[] memory newBalances = new uint256[](numTokens);
+        for (uint256 i = 0; i < numTokens; ++i) {
+            newBalances[i] = params.balancesScaled18[i];
+            if (i == params.indexIn) {
+                if (params.kind == SwapKind.EXACT_IN) {
+                    newBalances[i] += params.amountGivenScaled18;
+                } else {
+                    newBalances[i] += amountCalculatedScaled18;
+                }
+            } else if (i == params.indexOut) {
+                if (params.kind == SwapKind.EXACT_IN) {
+                    newBalances[i] -= amountCalculatedScaled18;
+                } else {
+                    newBalances[i] -= params.amountGivenScaled18;
+                }
+            }
+        }
+
+        uint256 newTotalImbalance = StableSurgeMedianMath.calculateImbalance(newBalances);
+
+        // If we are balanced, or the balance has improved, do not surge: simply return the regular fee percentage.
+        if (newTotalImbalance == 0) {
+            return staticFeePercentage;
+        }
+
+        uint256 oldTotalImbalance = StableSurgeMedianMath.calculateImbalance(params.balancesScaled18);
+
+        if (newTotalImbalance <= oldTotalImbalance || newTotalImbalance <= surgeThresholdPercentage) {
+            return staticFeePercentage;
+        }
+
+        // surgeFee = staticFee + (maxFee - staticFee) * (pctImbalance - pctThreshold) / (1 - pctThreshold).
```

Comment:
Can you expand this a bit further? E.g. explain how this behaves in the limits. When `pctImbalance --> 1`, `surgeFee = maxFee`, when `newTotalImbalance == surgeThresholdPercentage, surgeFee = staticFee` (case above).